juju/juju
# destroy-controller: destroy/release storage flags

## Description of change

Add --destroy-storage and --release-storage flags
to the destroy-controller command. If the user
specifies neither, and there is persistent storage
remaining, then the destruction fails and they
will be prompted to choose one of the methods of
storage removal.

The Controller API client and server facade have
both been updated accordingly. The Controller API
has been bumped to version 4. If a new client
calls destroy-controller against an older server,
they will be forced to use --destroy-storage, as
the old server does not complain if there is any
persistent storage remaining. If an old client
calls destroy-controller on the new controller,
it will destroy the storage as before.

## QA steps

(old server, new client)
1. juju bootstrap localhost (with juju 2.2.2)
2. juju destroy-controller -y localhost (with client from this branch)
(should fail, saying you have to use --destroy-storage)
3. juju destroy-controller -y localhost --destroy-storage
(should succeed)

(new server, old client)
1. juju bootstrap localhost && juju deploy postgresql --storage pgdata=lxd (with this branch)
2. juju destroy-controller -y localhost --destroy-all-models (with 2.2.2 client)
(should succeed and destroy storage)

(new server, new client)
1. juju bootstrap localhost && juju deploy postgresql --storage pgdata=lxd
2. juju destroy-controller -y localhost --destroy-all-models
(should fail, saying you must choose either --destroy-storage or --release-storage)
3. juju destroy-controller -y localhost --destroy-all-models --destroy-storage
(should succeed, and storage should be destroyed)

same as last scenario, but with --release-storage, and the LXD storage volume should not be destroyed

## Documentation changes

Yes, we need to document this new command behaviour.

## Bug reference

None.
